dvc/remote/base.py

```diff
@@ -339,63 +339,74 @@ def _process(
 
         with Tqdm(total=total, unit="file", desc=desc) as pbar:
             func = pbar.wrap_fn(func)
-            with ThreadPoolExecutor(max_workers=jobs) as executor:
-                if download:
-                    from_infos, to_infos, names, _ = (
-                        d + f for d, f in zip(dir_plans, file_plans)
-                    )
-                    fails = sum(
-                        executor.map(func, from_infos, to_infos, names)
-                    )
-                    if fails:
-                        self.index.clear()
-                        raise DownloadError(fails)
-                else:
-                    self._upload_plans(
-                        dir_plans,
-                        file_plans,
-                        dir_contents,
-                        missing_files,
-                        executor,
-                        jobs,
-                        func,
-                    )
+            self._process_plans(
+                download,
+                dir_plans,
+                file_plans,
+                dir_contents,
+                missing_files,
+                jobs,
+                func,
+            )
 
         return total
 
-    def _upload_plans(
+    def _process_plans(
```

Comment:
I felt like `_process` was actually doing so much, so extracted the actual transfer functionality aside.